lib/limiter/connlimiter.go

```diff
@@ -20,47 +20,64 @@ package limiter
 
 import (
 	"net/http"
+	"strings"
 	"sync"
 
-	"github.com/gravitational/oxy/connlimit"
-	"github.com/gravitational/oxy/utils"
 	"github.com/gravitational/trace"
-	log "github.com/sirupsen/logrus"
 )
 
-// ConnectionsLimiter is a network connection limiter and tracker
+// ConnectionsLimiter is a network connection limiter.
 type ConnectionsLimiter struct {
-	*connlimit.ConnLimiter
 	maxConnections int64
 
+	next http.Handler
+
 	sync.Mutex
 	connections map[string]int64
 }
 
 // NewConnectionsLimiter returns new connection limiter, in case if connection
 // limits are not set, they won't be tracked
-func NewConnectionsLimiter(config Config) (*ConnectionsLimiter, error) {
-	limiter := ConnectionsLimiter{
-		maxConnections: config.MaxConnections,
+func NewConnectionsLimiter(maxConnections int64) (*ConnectionsLimiter, error) {
+	// TODO: remove error from return type
+	return &ConnectionsLimiter{
+		maxConnections: maxConnections,
 		connections:    make(map[string]int64),
+	}, nil
+}
+
+// Wrap wraps an HTTP handler.
+func (l *ConnectionsLimiter) Wrap(h http.Handler) {
+	l.next = h
+}
+
+func (l *ConnectionsLimiter) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if l.next == nil {
+		sc := http.StatusInternalServerError
+		http.Error(w, http.StatusText(sc), sc)
+		return
 	}
 
-	ipExtractor, err := utils.NewExtractor("client.ip")
-	if err != nil {
-		return nil, trace.Wrap(err)
+	// TODO: use net.SplitHostPort to be more compatible with IPv6
+	vals := strings.SplitN(r.RemoteAddr, ":", 2)
+	if len(vals[0]) == 0 {
+		// TODO: log failed to extract source IP
+		sc := http.StatusInternalServerError
+		http.Error(w, http.StatusText(sc), sc)
+		return
 	}
 
-	limiter.ConnLimiter, err = connlimit.New(nil, ipExtractor, config.MaxConnections)
-	if err != nil {
-		return nil, trace.Wrap(err)
+	token := vals[0]
+	if err := l.AcquireConnection(token); err != nil {
+		// TODO: log "limiting request source %s: %v"
+		w.WriteHeader(http.StatusTooManyRequests)
+		w.Write([]byte(trace.UserMessage(err)))
```

Comment:
Will this break the error structure? 

From the API perspective, this is a breaking change that will disrupt the front-end error handling logic.

Namely, After this change the response will contain an unstructured text error message like:

"too many connections from ..."

Previously, the logic leveraged the HTTP error handling middleware to return a structured JSON response like
```
{
    "error": {
        "message": "too many connections from ..."
    }
}
```


To fix it in this case probably the[ trace.WriteError(w, err)](https://github.com/gravitational/trace/blob/master/httplib.go#L10) should be used instead of: 
```
w.WriteHeader(http.StatusTooManyRequests)
w.Write([]byte(trace.UserMessage(err)))
```

Reply:
Thanks!

Reply:
Are we sure the prior code didn't just write unstructured text?

Sure looks like it to me: https://github.com/gravitational/oxy/blob/f855322f2a6cd7219266bc12ac771f86b93f80f4/connlimit/connlimit.go#L120-L121

Reply:
Yeah, I can confirm we just send a text/plain response with a 429 status code.

`trace.WriteError` is probably better, but it would be a behavior change.

Reply:
All right then my comment is invalid - Sorry for confusion.